lsp-types/src/Language/LSP/Protocol/Message/Types.hs

```diff
@@ -0,0 +1,213 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TupleSections    #-}
+{-# LANGUAGE TypeInType       #-}
+
+module Language.LSP.Protocol.Message.Types where
+
+import           Language.LSP.Protocol.Types.Common
+import           Language.LSP.Protocol.Internal.Lens
+import           Language.LSP.Protocol.Internal.Method
+import           Language.LSP.Protocol.Types
+import           Language.LSP.Protocol.Message.LspId
+import           Language.LSP.Protocol.Message.Meta
+import           Language.LSP.Protocol.Message.Method ()
+import           Language.LSP.Protocol.Utils.Misc
+
+import           Control.Lens.TH
+import           Data.Aeson                         hiding (Null)
+import qualified Data.Aeson                         as J
+import           Data.Aeson.TH
+import           Data.Kind
+import           Data.String                        (IsString (..))
+import           Data.Text                          (Text)
+import           GHC.Generics
+import           GHC.TypeLits                       (KnownSymbol)
+
+-- 'RequestMessage', 'ResponseMessage', 'ResponseError', and 'NotificationMessage'
+-- aren't present in the metamodel, although they should be.
+-- https://github.com/microsoft/vscode-languageserver-node/issues/1079
+
+-- | Notification message type as defined in the spec.
+data NotificationMessage =
+  NotificationMessage
+    { _jsonrpc :: Text
+    , _method  :: Text
+    , _params  :: Maybe Value
+    } deriving stock (Show, Eq, Generic)
+
+deriveJSON lspOptions ''NotificationMessage
+
+-- This isn't present in the metamodel.
+-- | Request message type as defined in the spec.
+data RequestMessage = RequestMessage
+    { _jsonrpc :: Text
+    , _id      :: Int32 |? Text
+    , _method  :: Text
+    , _params  :: Maybe Value
+    } deriving stock (Show, Eq, Generic)
+
+deriveJSON lspOptions ''RequestMessage
+
+-- | Response error type as defined in the spec.
+data ResponseError =
+  ResponseError
+    { _code    :: ErrorCodes
```

Comment:
ResponseError takes ErrorCodes, but you have two types of error codes: ErrorCodes and LSPErrorCodes.